barretenberg/cpp/src/barretenberg/bb/main.cpp

```diff
@@ -364,9 +370,14 @@ void client_ivc_prove_output_all_msgpack(const std::string& bytecodePath,
     // TODO(https://github.com/AztecProtocol/barretenberg/issues/1116): remove manual setting of is_kernel once databus
     // has been integrated into noir kernel programs
     bool is_kernel = false;
-    for (Program& program : folding_stack) {
+    size_t stack_size = folding_stack.size();
+    for (size_t i = 0; i < stack_size; i++) {
+        auto program = folding_stack[i];
+        // Only the last circuit should be recursive.
+        bool is_last = i == stack_size - 1;
         // Construct a bberg circuit from the acir representation then accumulate it into the IVC
-        auto circuit = create_circuit<Builder>(program.constraints, 0, program.witness, false, ivc.goblin.op_queue);
+        auto circuit = create_circuit<Builder>(
+            program.constraints, recursive && is_last, 0, program.witness, false, ivc.goblin.op_queue);
```

Comment:
@vezenovm Previously when we had a program stack for folding which was marked `#[recursive]` only the final circuit would have been marked recursive. This seems quite odd to me and that we should use a consistent hash function across the full stack, do you have any insight into this?

Reply:
Folding is separate from the `#[recursive]` attribute as it is a different recursion scheme, and the recursion is going to be happening on one users device. The final circuit though is what will actually be sent to a prover in the network for which they need to recursively verify the folded stack of circuits. `#[recursive]` is only for when we want the traditional IVC scheme that passes an aggregation object with the proof.

Reply:
Lucas says it's all recursive, and that the `PrivateKernel*` circuits are supposed to be recursive as well. Also that UltraPlonk is the only one that cares about the flag at the moment, although UltraHonk might do so in the future.

Reply:
Yes folding is still a recursive scheme, so we can mark them all recursive for consistency but as Lucas mentioned UltraPlonk is the only one that cares about that flag so it seems that the final circuit only being marked `#[recursive]` was possibly just some technical debt as it is not being used with UH.